pkg/kubelet/prober/worker.go

```diff
@@ -230,7 +230,14 @@ func (w *worker) doProbe() (keepGoing bool) {
 	}
 
 	// Graceful shutdown of the pod.
-	if w.pod.ObjectMeta.DeletionTimestamp != nil && (w.probeType == liveness || w.probeType == startup) {
+	p, ok := w.probeManager.podManager.GetPodByUID(w.pod.UID)
```

Comment:
There's an overlapping change that needs to be made before this will be totally correct - you can't rely on the pod manager to correctly indicate whether a pod "has terminated", because the pod manager just tracks desired pods, not "pods the kubelet is running".

There is a separate change https://github.com/kubernetes/kubernetes/issues/105014 that probably needs to be done before this merges to more correctly take into account how pod lifecycle must handle probes.

However, even after that change this may not be the correct behavior, because a pod that has defined a liveness probe AND expects a long grace period may specify a shorter termination period for liveness failures (i.e. a container that may take 10 minutes to clean up is allowed to define a liveness probe that if fails, shuts the pod down in 10 seconds).  For more detail on that, see 

    kubectl explain deployment.spec.template.spec.containers.livenessProbe.terminationGracePeriodSeconds

Reply:
This change that relies on pod manager is only to indicate whether a pod "has deleted",  not "has terminated".
As the housekeeping still would shutdown probes for undesired pods, https://github.com/kubernetes/kubernetes/blob/40c2d049465417f510e4182b05953a49fc5693d4/pkg/kubelet/kubelet_pods.go#L1106

Actually if we consider the feature ProbeTerminationGracePeriod, this logic above for probe is not correct, I just want to fix that bug https://github.com/kubernetes/kubernetes/issues/107473 at that time. Now I think these codes may be not useful, actually from our tests, the probes for terminating(deteted) pod are basically always shutted down by housekeeping, the codes above have no chance to run.